### ISSUE TITLE
[New issue]: Training-Validation curves on Tensorboard 

### ISSUE THREAD

**Issue by manos-mark** [fedbiomed/fedbiomed#914]
An issue raised while I was trying to plot training curves with Tensorboard on training data using the `training_step` abstract function of the training plan. In contrast with `testing_step` where we can calculate custom metrics, and without any further implementation those metrics are plotted on Tensorboard, using `training_step` and training data this is not possible. 

Following the reporting of this issue on the [Discord support chat](https://discord.com/channels/927917604188852235/927919698862047292/1173932632480948234), @sharkovsky recommends the following to handle this problem: 

> We need to redefine the _train_over_batch and training_routine functions of the TorchTrainingPlan.  The idea would be to change the training routine such that it can accept additional loss metrics, and then copy the code for using the history monitor to plot them. 
> 
> You would need to (in order of how things are executed):
> 1) copy and paste the code for these two functions from TorchTrainingPlan and redefine them in your own TrainingPlan. You can find the source code in the file _torchnn.py.
> 
> 2) change the training routine to accept the computed metrics  (the equivalent of line 414 of the _torchnn.py file) from corrected_loss, loss = self._train_over_batch(data, target) to corrected_loss, loss, metrics = self._train_over_batch(data, target)
> 
> 3) change the train over batch function to accept the computed metrics (the equivalent of line 499 of the _torchnn.py file) from loss = self.training_step(data, target) to loss, metrics = self.training_step(data, target). 
> 
> 4) change your definition of training_step to return your custom metrics in addition to the loss value
> 
> 5) change the train over batch function to return the custom metrics (the equivalent of line 519 of the _torchnn.py file)  from return corrected_loss, loss  to return corrected_loss, loss, metrics
> 
> 6) add code to training routine for logging the metrics using the history monitor. Your mileage may vary but it would be something like 
> 
> history_monitor.add_scalar(metric={'MyMetric': metrics.item()},
> iteration=num_iter,
> epoch=epoch_to_report,
> train=True,
> num_samples_trained=num_samples,
> num_batches=num_iter_max,
> total_samples=num_samples_max,                  batch_samples=batch_size)
> 
> for point 6, you can take inspiration from line 436 of `_torchnn.py`
> https://github.com/fedbiomed/fedbiomed/blob/dffcf018f0a6a38e8ea1999b7c1a75deb8fe1c60/fedbiomed/common/training_plans/_torchnn.py#L436
> 
> but note that if you put your code inside the same indentation as line 436, you will only receive a report every `log_interval` iterations. If you want to log every batch, I suggest setting `log_interval=1` in your `training_args`.

**Comment by manos-mark** [comment]
Thank you very much @sharkovsky, your solution works pretty well! I comment on some minor modifications that you need to add to YourTrainingPlan class:

1. Add imports to your notebook because you copy two functions from `_torchnn.py`
```
from typing import Any, Dict, List, Tuple, OrderedDict, Optional, Union, Iterator
from fedbiomed.common.training_plans._training_iterations import MiniBatchTrainingIterationsAccountant
from fedbiomed.common.logger import logger
from fedbiomed.common.training_plans._base_training_plan import BaseTrainingPlan

ModelInputType = Union[torch.Tensor, Dict, List, Tuple]
```

2. Add those imports to your training plan `init_dependencies` function 
```
def init_dependencies(self):
    deps = [
        ...           
        'from typing import Any, Dict, List, Tuple, OrderedDict, Optional, Union, Iterator',
        'ModelInputType = Union[torch.Tensor, Dict, List, Tuple]',
        'from fedbiomed.common.training_plans._training_iterations import MiniBatchTrainingIterationsAccountant',
        'from fedbiomed.common.logger import logger',
        'from fedbiomed.common.training_plans._base_training_plan import BaseTrainingPlan',
    ]
    return deps
```
3. Concerning the step 6 from the initial issue,  I added code to the `training_routine` function inside the if statement that checks `if history_monitor is not None`, to iterate through my custom metrics dictionary using `_create_metric_result_dict` from the `BaseTrainingPlan` class:
```
for metric_name, metric_value in metrics.items():
    m_dict = BaseTrainingPlan._create_metric_result_dict(metric=metric_value, metric_name=metric_name)
    history_monitor.add_scalar(
        metric=m_dict,
        iteration=num_iter,
        epoch=epoch_to_report,
        train=True,
        num_samples_trained=num_samples,
        num_batches=num_iter_max,
        total_samples=num_samples_max, batch_samples=batch_size
    )
```

**Comment by sharkovsky** [comment]
Amazing work @manos-mark , thank you for getting back to us! 
it was not an easy customization, congrats on making it! 

Do you think we can close the issue?

**Comment by manos-mark** [comment]
Thanks for your kind words, but you provided all the instructions :smile: 

From my point of view, the issue is resolved, but there is still space for improvement. Right now, the training metrics are plotted per batch but the validation metrics are plotted per global epoch. So, the training metrics on each batch should be averaged to be able to plot them next to the validation metrics. I am not sure if my explanation is very clear, so I will try to implement the code soon and provide it in the following comments.

**Comment by sharkovsky** [comment]
Yeah, there is a lot of space for improvement for our metrics reporting in general...
I will close this issue for now, but I'll be happy to receive a pull request from you in case you want to extend/improve Fed-BioMed's reporting some time in the future 😄 😄
